Files with only comments are no longer always stale

jgmchan opened the pull request:
At the beginning of compilation, initial `:source` records of the files that
needs to be compiled are added to an Agent process. These records are then
updated when the files are compiled and then written to the compile manifest.

The `:size` field in the initial record has a default of 0, and since files
without any code (they only have comments or newlines) are not compiled, this
is never updated with the actual size of the file before the compile manifest
is written to disk.

These files are always seen as stale because their size on disk does not match
the compile manifest.

This is fixed by adding the actual size of the file to the initial record.

Closes #7066

josevalim commented on lib/mix/lib/mix/compilers/elixir.ex:
```diff
@@ -334,7 +334,11 @@ defmodule Mix.Compilers.Elixir do
+    Enum.reduce(
+      changed,
+      sources,
+      &List.keystore(&2, &1, source(:source), source(source: &1, size: :filelib.file_size(&1)))
```
We already have this information from a couple lines above. I would prefer to re-use it instead of reaching the file system again.

jgmchan replied:
👍

josevalim commented on lib/mix/test/mix/tasks/compile_test.exs:
```diff
@@ -34,6 +34,25 @@ defmodule Mix.Tasks.CompileTest do
     assert Mix.Tasks.Compile.manifests() |> Enum.map(&Path.basename/1) == [".compile.elixir"]
   end
 
+  test "compile a project with files containing only comments or newlines" do
```
Please see "does not recompile empty files" in compile.elixir_test.exs. I think we can write the same test, just add some comments.

jgmchan replied:
Ah cool, I'll use that test instead.

jgmchan commented:
Updated PR addressing the comments. It now uses the information we already had previously without going to the filesystem again. Please let me know if the code can be improved 😄

jgmchan commented on lib/mix/lib/mix/compilers/elixir.ex:
```diff
@@ -267,10 +275,12 @@ defmodule Mix.Compilers.Elixir do
       source(
         source: source,
-        size: :filelib.file_size(source),
+        size: size,
```
Since we now have the correct file size information inside the Process dictionary, I think we probably don't need to access the file system here anymore.

josevalim commented on lib/mix/lib/mix/compilers/elixir.ex:
```diff
@@ -45,27 +45,31 @@ defmodule Mix.Compilers.Elixir do
         # changed, let's just compile everything
-        MapSet.to_list(all_paths)
+        all_paths = MapSet.to_list(all_paths)
+        {all_paths, mtimes_and_sizes(all_paths)}
```
Let's not call mtimes_and_sizes here. Instead let's do a simple comprehension:

    for path <- all_paths, into: %{}, do: {path, Mix.Utils.last_modified_and_size(path)}

We don't need to worry about dupes here and this way we don't need to change mtimes_and_sizes.

josevalim commented on lib/mix/lib/mix/compilers/elixir.ex:
```diff
@@ -332,9 +342,28 @@ defmodule Mix.Compilers.Elixir do
   ## Resolution
 
-  defp update_stale_sources(sources, changed) do
+  defp update_stale_sources(sources, changed, sources_stats \\ %{}) do
```
It would be better to have two functions. One that receives the map and uses the first clause of the cond and another function that does exclusively the second clause of the cond.

josevalim commented on lib/mix/lib/mix/compilers/elixir.ex:
```diff
@@ -332,9 +342,28 @@ defmodule Mix.Compilers.Elixir do
+          source(size: size) = source
+          size
+
+        true ->
```
Do we ever hit this clause? If so, when? Why? :)

jgmchan replied:
Yeah we shouldn't ever hit this clause and it should fail if it ever does :)

josevalim commented:
I love the direction this is going @jgmchan! I have added a couple more comments.

jgmchan commented:
Ok, updated the PR with suggestions from the comments. Please let me know how it looks.

josevalim commented:
If you are interested, I further clean up the code here: 2ea726e15bace29e76f58d8eda1381db56cce190
